clang/lib/AST/PreorderAST.cpp

```diff
@@ -56,11 +62,58 @@ void PreorderAST::Create(Expr *E, Node *Parent) {
   E = Lex.IgnoreValuePreservingOperations(Ctx, E->IgnoreParens());
 
   if (const auto *BO = dyn_cast<BinaryOperator>(E)) {
-    auto *N = new BinaryNode(BO->getOpcode(), Parent);
+    BinaryOperator::Opcode BinOp = BO->getOpcode();
+    Expr *LHS = BO->getLHS();
+    Expr *RHS = BO->getRHS();
+
+    // We can convert (e1 - e2) to (e1 + -e2) if -e2 does not overflow.  One
+    // instance where -e2 can overflow is if e2 is INT_MIN. Here, instead of
+    // specifically checking whether e2 is INT_MIN, we add a unary minus to e2
+    // and then check if the resultant expression -e2 overflows. If it
+    // overflows, we undo the unary minus operator.
+
+    // TODO: Currently we only handle detection of overflow for constant
+    // expressions. We need to handle the case where e2 is a variable
+    // expression which is not known until run time.
+    if (BO->getOpcode() == BO_Sub) {
+      RHS = new (Ctx) UnaryOperator(RHS, UO_Minus, RHS->getType(),
+                                    RHS->getValueKind(),
+                                    RHS->getObjectKind(),
+                                    SourceLocation(),
+                                    /*CanOverflow*/ true);
+      bool Overflow = false;
+      if (RHS->isIntegerConstantExpr(Ctx)) {
+        SmallVector<PartialDiagnosticAt, 8> Diag;
+        llvm::APSInt IntVal =
+          RHS->EvaluateKnownConstIntCheckOverflow(Ctx, &Diag);
+
+        for (auto &PD : Diag) {
+          if (PD.second.getDiagID() == diag::note_constexpr_overflow) {
+            Overflow = true;
+            break;
+          }
+        }
+      }
+
+      if (Overflow)
+        RHS = dyn_cast<UnaryOperator>(RHS)->getSubExpr();
+      else
+        BinOp = BO_Add;
+    }
+
+    auto *N = new BinaryNode(BinOp, Parent);
     AddNode(N, Parent);
 
-    Create(BO->getLHS(), /*Parent*/ N);
-    Create(BO->getRHS(), /*Parent*/ N);
+    Create(LHS, /*Parent*/ N);
+    Create(RHS, /*Parent*/ N);
+
+  } else if (!Parent) {
+    // We will enter here for expressions like "p" which are not BinaryNode and
+    // which do not yet have a parent. So we create a parent BinaryNode with
+    // addition as the operator and add "p" as a LeafExprNode of this node.
+    auto *N = new BinaryNode(BO_Add, Parent);
```

Comment:
Why is this invariant necessary?  The invariant seems to be that the root node must be a BinaryNode.  I'm not following why it is needed.

Reply:
Yes, the invariant is that the root node must be a BinaryNode. So for expressions like `if (*p)`, we don't have a BinaryOperator. So when we enter the Create function there is no root and the parent is null. So in this condition we create a new BinaryNode with + as the operator and add p as a child of this BinaryNode. Later in NormalizeExprsWithoutConst we normalize p to p+0 by adding 0 as a sibling of p.